src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Publisher.kt

```diff
@@ -22,54 +22,47 @@ import com.google.cloud.pubsub.v1.Publisher
 import com.google.protobuf.ByteString
 import com.google.pubsub.v1.PubsubMessage
 import com.google.pubsub.v1.TopicName
-import java.io.ByteArrayOutputStream
 
-fun createMessage(message: String): PubsubMessage {
-    val data = ByteString.copyFromUtf8(message)
-
-    return PubsubMessage.newBuilder()
-            .setData(data)
+fun getResultData(taskId: String, statusCode: Int, resultBytes: ByteArray): ByteString {
+    return CompilerFleet.CompilerFleetResult.newBuilder()
+            .setTaskId(taskId)
+            .setStatusCode(statusCode)
+            .setResultBytes(ByteString.copyFrom(resultBytes))
             .build()
+            .toByteString()
 }
 
 class Publisher(private val topicName: String) {
-    private val PROJECT_ID = ServiceOptions.getDefaultProjectId()
+    private val publishManager: Publisher
 
-    fun publish(zipBytes: ByteArray, rootFileName: String, taskId: String) {
+    init {
+        val project_id = ServiceOptions.getDefaultProjectId()
         val topicId = this.topicName
-        val serviceTopicName = TopicName.of(this.PROJECT_ID, topicId)
-        var publisher: Publisher? = null
+        val serviceTopicName = TopicName.of(project_id, topicId)
 
-        try {
-            publisher = Publisher.newBuilder(serviceTopicName).build()
-            val byteOutput = ByteArrayOutputStream()
-            CompilerFleet.CompilerFleetRequest.newBuilder()
-                    .setZipBytes(ByteString.copyFrom(zipBytes))
-                    .setRootFileName(rootFileName)
-                    .setTaskId(taskId)
-                    .build()
-                    .writeTo(byteOutput)
+        this.publishManager = Publisher.newBuilder(serviceTopicName).build()
+    }
 
-            val data = ByteString.copyFrom(byteOutput.toByteArray())
+    fun publish(data: ByteString) {
+        try {
             val pubsubMessage = PubsubMessage.newBuilder()
                     .setData(data)
                     .build()
 
-            val future = publisher.publish(pubsubMessage)
+            val future = publishManager.publish(pubsubMessage)
 
             ApiFutures.addCallback(future, object : ApiFutureCallback<String> {
 
                 override fun onFailure(throwable: Throwable) {
                     println(throwable)
-                    println("Error publishing taskID : $taskId")
                 }
 
                 override fun onSuccess(messageId: String) {
                     println("successful published $messageId")
                 }
             })
         } finally {
-            publisher?.shutdown()
+            publishManager.shutdown()
```

Comment:
Removed finally block. Do I need to catch exception or just leave it to command line?

Reply:
It is called not only from the (not yet existing) command line client, but also from TaskReceiver. I don't think that we should care about the excetions thrown from `publish` method. More interesting is `onFailure` in the callback which in general should be processed somehow (e.g. by retrying or by sending failure message) but it might be out of scope of this pull request.